photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/DistributedGLMLossFunction.scala

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function.glm
+
+import breeze.linalg._
+import org.apache.spark.SparkContext
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.RDD
+
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.function._
+import com.linkedin.photon.ml.normalization.NormalizationContext
+import com.linkedin.photon.ml.optimization.{GLMOptimizationConfiguration, RegularizationType}
+
+/**
+ * This class is used to calculate the value, gradient, and Hessian of generalized linear models for distributed
+ * optimization problems. The loss function of a generalized linear model can all be expressed as:
+ *
+ * L(w) = \sum_i l(z_i, y_i)
+ *
+ * with:
+ *
+ * z_i = w^T^ z_i.
+ *
+ * Different generalized linear models will have different l(z, y). The functionality of l(z, y) is provided by a
+ * [[PointwiseLossFunction]]. Since the loss function could change for different types of normalization, a normalization
+ * context object indicates which normalization strategy is used to evaluate the loss function.
+ *
+ * @param singleLossFunction A single loss function l(z, y) used for the generalized linear model
+ * @param sparkContext The Spark context
+ * @param treeAggregateDepth The depth used by treeAggregate. Depth 1 indicates normal linear aggregate. Using
+ *                           depth > 1 can reduce memory consumption in the Driver and may also speed up the
+ *                           aggregation. It is experimental currently because treeAggregate is unstable in Spark
+ *                           versions 1.4 and 1.5.
+ */
+protected[ml] class DistributedGLMLossFunction private (
```

Comment:
Old GeneralizedLinearLossFunction class split into two variants (just like the optimization problems and objective functions).